solr/solrj/src/java/org/apache/solr/client/solrj/impl/HttpSolrJdkClient.java

```diff
@@ -278,64 +285,74 @@ private HttpResponse<InputStream> doPutOrPost(
-   *    the Jetty Server underpinning Solr does not accept this.
+   * <p>The JDK Http Client will send an upgrade request over Http/1 along with request content in
+   * the same request. However, the Jetty Server underpinning Solr does not accept this.
    *
-   *    By sending a ping before any real requests occur, the client
-   *    knows if Solr can accept Http/2, and no additional
-   *    upgrade requests will be sent.
+   * <p>We send a HEAD request first, then the client knows if Solr can accept Http/2, and no
+   * additional upgrade requests will be sent.
    *
-   *    See https://bugs.openjdk.org/browse/JDK-8287589
-   *    See https://github.com/jetty/jetty.project/issues/9998#issuecomment-1614216870
+   * <p>See https://bugs.openjdk.org/browse/JDK-8287589 See
+   * https://github.com/jetty/jetty.project/issues/9998#issuecomment-1614216870
    *
-   *    We only try once, and if it fails, we downgrade to Http/1
+   * <p>We only try once, and if it fails, we downgrade to Http/1
    *
    * @param url the url with no request parameters
    * @return true if success
    */
-  private synchronized boolean maybeTryHeadRequest(String url) {
-    if(headRequested) {
+  private boolean maybeTryHeadRequest(String url) {
+    if (forceHttp11 || url == null || url.toLowerCase(Locale.ROOT).startsWith("https://")) {
+      return true;
+    }
+    return maybeTryHeadRequestSync(url);
+  }
+
+  protected volatile boolean headRequested; // must be threadsafe
+  private boolean headSucceeded; // must be threadsafe
+
+  private synchronized boolean maybeTryHeadRequestSync(String url) {
+    if (headRequested) {
       return headSucceeded;
     }
+
     URI uriNoQueryParams;
     try {
       uriNoQueryParams = new URI(url);
-    } catch(URISyntaxException e) {
-      headSucceeded = false;
+    } catch (URISyntaxException e) {
+
+      // If the url is invalid, let a subsequent request try again.
       return false;
     }
-    HttpRequest.Builder headReqB = HttpRequest.newBuilder(uriNoQueryParams).method("HEAD", HttpRequest.BodyPublishers.noBody());
+    HttpRequest.Builder headReqB =
+        HttpRequest.newBuilder(uriNoQueryParams)
+            .method("HEAD", HttpRequest.BodyPublishers.noBody());
     decorateRequest(headReqB, new QueryRequest());
     try {
       httpClient.send(headReqB.build(), HttpResponse.BodyHandlers.discarding());
       headSucceeded = true;
-    } catch(IOException ioe) {
+    } catch (IOException ioe) {
       log.warn("Could not issue HEAD request to {} ", url, ioe);
       headSucceeded = false;
-    } catch(InterruptedException ie) {
+    } catch (InterruptedException ie) {
       Thread.currentThread().interrupt();
       headSucceeded = false;
     } finally {
+
+      // The HEAD request is tried only once.  All future requests will skip this check.
       headRequested = true;
+
+      if (!headSucceeded) {
+        log.info("All insecure POST requests with a chunked body will use http/1.1");
```

Comment:
Would replacing "insecure" with "HTTP" or "unencrypted" or "non-SSL" be fine?  To me, there's something about the word "insecure" that refers to overall security posture, thus implying the user is negligent here which is not necessarily the case.

Reply:
I can change this to _unencrypted_.  Let's try not to offend our users, eh?